Use bytesrepr (de)serialization in the storage component

georgepisaltu opened the pull request:
Partially fixes: #2908

This PR adds bytesrepr (de)serialization support for:
- `BlockHeaderWithMetadata`
- `BlockSignatures`
- `BlockWithMetadata`
- `FinalizedApprovals`
- `FinalizedApprovalsWithId`
- `DeployMetadata`
- `TrieOrChunkId`
- `GossipedAddress`

Additionally, `bincode` (de)serialization was removed from the storage component and replaced by `bytesrepr` with the exception of `decode_item_id`, which wasn't changed yet because it's not used only by the storage component.

HansWestman74 committed:
Update CHANGELOG.md

Removed git leftovers

fizyk20 committed:
Move time types to casper_types

fizyk20 committed:
Appease Clippy

fizyk20 committed:
Remove an unused dependency

georgepisaltu committed:
Add bytesrepr support for TrieOrChunkId

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Add bytesrepr support for GossipedAddress

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Add bytesrepr support for DeployMetadata

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Add bytesrepr support for FinalizedApprovals

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Add bytesrepr support for FinalizedApprovalsWithId

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Added tests for new bytesrepr impls in types/deploy.rs

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Add bytesrepr support for BlockSignatures

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Add bytesrepr support for BlockHeaderWithMetadata

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Add bytesrepr support for BlockWithMetadata

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Added tests for new bytesrepr impls in types/block.rs

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Add support for bytesrepr (de)ser in storage

This commit adds bytesrepr serialization/deserialization support for
structures used in the storage component. The current serde/bincode
support is not yet removed.

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

TomVasile committed:
use ci.json file for default protoco1 selection

rafal-ch commented on node/src/components/small_network/gossiped_address.rs:
```diff
@@ -28,6 +50,31 @@ impl Display for GossipedAddress {
+impl FromBytes for GossipedAddress {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), BytesreprError> {
+        let (sock_addr_str, bytes): (String, &[u8]) = FromBytes::from_bytes(bytes)?;
+        let sock_addr = std::str::FromStr::from_str(sock_addr_str.as_str())
```
```suggestion
        let sock_addr: SocketAddr = sock_addr_str.parse().map_err(|_| BytesreprError::Formatting)?;
```

georgepisaltu replied:
Done.

rafal-ch commented on node/src/types/deploy.rs:
```diff
@@ -1614,6 +1677,69 @@ pub struct DeployMetadata {
+// bytesrepr is no-std and `HashMap` is only available in `std`.
+impl FromBytes for DeployMetadata {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), bytesrepr::Error> {
+        let (num_keys, mut stream) = u32::from_bytes(bytes)?;
```
Please consider using the `usize` as a type to represent number of keys. `usize` is what's returned from `HashMap::len()`.

georgepisaltu replied:
It seems `bytesrepr` (de)serialization of collection objects such as maps require the length to fit in an `u32`, hence the error `bytesrepr::Error::NotRepresentable`. As discussed, we'll keep the implementation consistent with the other code.

rafal-ch reviewed (approved):
LGTM with minor nits :+1:

fizyk20 committed:
Address review comments

georgepisaltu committed:
Use `FromStr::parse` in `GossipedAddress`

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

Fraser999 commented on node/src/components/small_network/gossiped_address.rs:
```diff
@@ -28,6 +50,32 @@ impl Display for GossipedAddress {
+    }
+
+    fn write_bytes(&self, writer: &mut Vec<u8>) -> Result<(), BytesreprError> {
+        self.0.to_string().write_bytes(writer)
```
Hmm - this is a bit suboptimal.  The string representation is likely to be more expensive in terms of computation and space.

I think it would be better to implement bytesrepr traits for the following std types in the bytesrepr module of casper-types:
* `Ipv4Addr`
* `Ipv6Addr`
* `IpAddr`
* `SocketAddrV4`
* `SocketAddrV6`
* `SocketAddr`

We should use the [`octets()`](https://doc.rust-lang.org/std/net/struct.Ipv6Addr.html#method.octets) method on `Ipv4Addr` and `Ipv6Addr` to be efficient.

We should probably share a couple of const tags for the implementation of the `IpAddr` and `SocketAddr` enums - something like:
```rust
pub const IP_V4_TAG: u8 = 0;
pub const IP_V6_TAG: u8 = 1;
```

georgepisaltu replied:
Done.

Fraser999 commented on execution_engine/src/storage/trie/mod.rs:
```diff
@@ -450,6 +450,31 @@ impl Display for TrieOrChunkId {
+    }
+
+    fn serialized_length(&self) -> usize {
+        ToBytes::serialized_length(&self.0) + ToBytes::serialized_length(&self.1)
```
It's a bit more canonical to write `self.0.serialized_length()`.  Same comment applies to a few such instances in this PR.

georgepisaltu replied:
Done.

Fraser999 commented on execution_engine/src/storage/trie/mod.rs:
```diff
@@ -450,6 +450,31 @@ impl Display for TrieOrChunkId {
+
+impl FromBytes for TrieOrChunkId {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), bytesrepr::Error> {
+        let (idx, bytes) = FromBytes::from_bytes(bytes)?;
```
I tend to favour being explicit about the type in cases like this - mainly for readability: `u64::from_bytes(bytes)`.  Same comment for a few instances in this PR.

georgepisaltu replied:
Done.

Fraser999 commented on node/src/types/deploy.rs:
```diff
@@ -1614,6 +1677,69 @@ pub struct DeployMetadata {
+    }
+}
+
+// `bytesrepr::FromBytes` is not implemented for the inner `HashMap` because
+// bytesrepr is no-std and `HashMap` is only available in `std`.
```
I think we should implement the traits for `HashMap` in casper-types if the `std` feature is enabled in that crate.

georgepisaltu replied:
Done.

mpapierski commented on node/src/types/deploy.rs:
```diff
@@ -1614,6 +1677,69 @@ pub struct DeployMetadata {
+            .try_into()
+            .map_err(|_| BytesreprError::NotRepresentable)?;
+        num_keys.write_bytes(writer)?;
+        for (key, value) in self.execution_results.iter() {
```
Can we sort the keys first? HashMap is unordered (and in stdlib hashmap specifically order is randomized) and two instances with the same data would generate different bytes. We should sort the keys first (or create a temporary `BTreeMap<&K, &V>` or smth) to make the representation deterministic.

georgepisaltu replied:
> two instances with the same data would generate different bytes

Is this a problem as long as deserializing those bytes yields the same result? Is there a use for the intermediary bytes representation?

dwerner replied:
@mpapierski, are you concerned about data ending up in the trie at some point? In that case, having a different order to the bytes would affect hashing, as well as extra duplication of mutated data over time. Ideally this would be handled in the ToBytes impl for HashMap that @Fraser999 is suggesting.

georgepisaltu replied:
I'll wait for an answer here to push the change.

mpapierski replied:
@dwerner, yes, exactly this is my concern - @georgepisaltu, we should enforce an order of keys in `ToBytes for HashMap`

georgepisaltu replied:
@mpapierski done.

dwerner reviewed (approved):
Looks good to me, approval pending @Fraser999 's suggestions being implemented.

georgepisaltu committed:
Add support for `std::net` structs in bytesrepr

Added bytesrepr support for:
- Ipv4Addr
- Ipv6Addr
- IpAddr
- SocketAddrV4
- SocketAddrV6
- SocketAddr

Bytesrepr implementations are only available with the `std` feature.

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Use bytesrepr::SocketAddr impl for GossipedAddress

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Make bytesrepr function calls more explicit

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Add support for HashMap in bytesrepr

Bytesrepr implementation is only available with the `std` feature.

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Use bytesrepr::HashMap impl for DeployMetadata

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

marc-casperlabs commented on .drone.yml:
```diff
@@ -138,11 +138,6 @@ steps:
 
 - name: nctl-upgrade-test
   <<: *buildenv
-  environment:
```
Is this accidental or on purpose? I am out of the loop here :)

Looks like some branching shenanigans going on?

georgepisaltu replied:
I did a `git merge` with upstream/dev because we needed the change which undeprecated the `std` feature gate on casper-types. I don't know exactly why it's showing up like this and not a single merge commit.

marc-casperlabs commented on node/src/components/storage.rs:
```diff
@@ -1953,6 +1952,10 @@ fn decode_item_id<T>(raw: &[u8]) -> Result<T::Id, GetRequestError>
+    // other parts of the code which we don't want to change right now. In order to
+    // fully migrate from bincode, we would have to use
+    // `bytesrepr::deserialize_from_slice` here.
     bincode::deserialize(raw).map_err(GetRequestError::MalformedIncomingItemId)
```
Yes, ItemIDs are bincode and this is shared across networked components, like the trie-store. Unfortunately, the IDs are raw `Vec<u8>` without a type wrapper at the moment, and have no serialization methods. Sorry.

georgepisaltu replied:
The next step is to replace `bincode` with `bytesrepr` in the small network component. After that is done, I think we could change this as well.

marc-casperlabs replied:
I think the networking component is not at fault here, but the fetcher responses and their lack of typing. I've been sitting on #2163 for too long, which should in theory fix this.

marc-casperlabs commented on types/src/bytesrepr.rs:
```diff
@@ -670,6 +684,56 @@ where
+{
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), Error> {
+        let (num_keys, mut stream) = u32::from_bytes(bytes)?;
+        let mut result = HashMap::new();
```
We should use `with_capacity(usize::from(num_keys))` here, I think.

georgepisaltu replied:
Done.

`usize` has `TryFrom` instead of `From` for `u32`. Seems like 16-bit systems are still a thing.

mpapierski replied:
I'd be careful here with `with_capacity` @marc-casperlabs - `num_keys` is potentially untrusted data, and if somehow the host tries to deserialize `u32::MAX` here on the host, it will abort the process. In `ToBytes for Vec<T>,` we use `try_vec_with_capacity` to avoid this problem.

marc-casperlabs replied:
> usize has TryFrom instead of From for u32. Seems like 16-bit systems are still a thing.

I specifically discounted them, I don't see us supporting those any time soon. But I guess it can't hurt _much_ to use `TryFrom` and we'll be happy to have done that once we port everything to the SNES ❤️ .

@mpapierski makes a very good point that I missed though --- I forgot that there _is_ a limit since the incoming chunk of raw serialized data is likely size limited (otherwise, adding infinite items would have had the same effect). So I guess no `with_capacity` (or a compromise using a small upper bound, but at that point, we might as well not bother, since the size of the key type might affect this as well).

tl;dr my bad, my suggestions weren't exactly stellar.

georgepisaltu replied:
Reverted the commit with the change.

marc-casperlabs commented on types/src/bytesrepr.rs:
```diff
@@ -772,6 +836,213 @@ impl<T: FromBytes, E: FromBytes> FromBytes for Result<T, E> {
+    }
+
+    fn write_bytes(&self, writer: &mut Vec<u8>) -> Result<(), Error> {
+        self.port().write_bytes(writer)?;
```
Is there a reason we are writing these in what, at least to me, is "reverse" order? The implementation is platform specific of course, but I assume most people expect these to be in "ip, port" order?

georgepisaltu replied:
No reason other than my preference of serializing primitives first. This only makes a difference in the intermediary bytes representation, which I'm still not sure it's [relevant](https://github.com/casper-network/casper-node/pull/2922#discussion_r864859094).

marc-casperlabs replied:
I usually view these things with rose colored glasses of us eventually getting to use a derive macro for these things. If that were the case, I would stick to the field order declared in struct itself, which defines at least one possible canonical ordering.

georgepisaltu replied:
I refactored it.

marc-casperlabs commented on types/src/bytesrepr.rs:
```diff
@@ -772,6 +836,213 @@ impl<T: FromBytes, E: FromBytes> FromBytes for Result<T, E> {
+    fn write_bytes(&self, writer: &mut Vec<u8>) -> Result<(), Error> {
+        self.port().write_bytes(writer)?;
+        self.flowinfo().write_bytes(writer)?;
+        self.scope_id().write_bytes(writer)?;
```
I am unsure whether we want to persist the `scope_id` and `flowinfo` parts. On the one hand, they are part of the serialization and naively I would expect for them to roundtrip correctly, on the other hand their presence on an IPv6 address seems more like an implementation detail of the underlying C struct.

`scope_id` is only valid on the local machine, as it names a local interface. Should we ever support IPv6, we probably would not be sending this across the network as a gossiped address.

`flowinfo` has meaning outside the local machine, but I wonder whether it should be persisted as well. If used, it is typically filled before sending.

My limited intuition would also opt for an "ip, port" serialization. Maybe someone with more comprehensive IPv6 experience can help out here?

georgepisaltu replied:
My personal preference here would be to keep `bytesrepr` implementations as strictly (de)serializing structures completely (no losses) and not actually reason about what fields are relevant or not in the implementation. I think that kind of reasoning should be done at a higher level, kind of how `serde` does it with `bincode`. This is to keep the `bytesrepr` functionality consistent and to avoid any weird corner cases in its behavior. In the absence of a higher level implementation, you would need to either:
- Create new structs which serve as a schema containing exactly what you need (de)serialized. In this particular case, we would have a
```
CustomSocketAddrV6 {
    ip: IpV6Addr,
    port: u16,
}
```
- Reason about the fields pre-serialization/post-deserialization.

marc-casperlabs replied:
That's something I can certainly support. I kinda blame the actual types tacking on these extra fields :/

My reasoning is that should we ever use a serialized representation of this in a hashed way, we'd have to zero some of the fields.

Right now, the only thing I can think of Ipv6 addresses being used for is them being sent to other nodes as part of the address gossiping the network component does, should we implement IPv6 support. Now in the past we once had signed addresses being sent to peers, and it would have been weird for the gossipped address to local-only information, which it then would have to send on unchanged to keep the signature intact.

It's certainly not the end of the world though, we'd just have to implement our own `SocketAddr` type and create a `From`-impl.

So, since I can't give a really good reason to not serialize it, let's go with your solution :)

marc-casperlabs reviewed (commented):
Some open questions, but I would not be overly upset if someone ignored my comments and pressed on with it :)

georgepisaltu committed:
Avoid reallocations in bytesrepr::HashMap

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

marc-casperlabs commented:
LGTM, sans my ill-guided suggestion that @mpapierski pointed out and will need to be fixed.

georgepisaltu committed:
Revert "Avoid reallocations in bytesrepr::HashMap"

This reverts commit 4ce59f4001a31a497d3491c629c6c7beaac16566.

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Rearrange (de)ser order for net structs in bytesrepr

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Make HashMap bytesrepr (de)ser deterministic

Signed-off-by: George Pisaltu <georgep@casperlabs.io>
